### PR TITLE
Load inputs from external configuration files in Elastic Agent in standalone mode

### PR DESCRIPTION
**Pull request by kvch**
## What does this PR do?

This PR adds support for loading external configuration files to load inputs when running Agent in standalone mode.
Users have to put their configuration files into `{path.config}/inputs.d` with the extension `yml`.

When it is configured, Agent takes the `inputs` configuration from the last standalone configuration and appends the inputs from the external configuration files to it. When the standalone configurations are reloaded, external input configurations are reloaded as well.

The external configuration file must include `inputs`, otherwise the file is invalid. Example for correct file:

```yaml
inputs:
- data_stream:
    dataset: system.auth
    type: logs
  id: logfile-system.auth-my-id
  paths:
  - /var/log/auth.log*
  use_output: default
```

## Why is it important?

Users can configure external configuration files, similar to Filebeat. So managing configuration can be done by Git or Chef instead of Fleet.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: x-pack/elastic-agent/_meta/config/common.reference.p2.yml.tmpl
Comment (ruflin) [inline x-pack/elastic-agent/_meta/config/common.reference.p2.yml.tmpl:138]: Do we need to make the path configurable? Can't we just rely on a predefined directory an pattern?

Comment (kvch) [reply x-pack/elastic-agent/_meta/config/common.reference.p2.yml.tmpl:138]: It is possible. I chose to make it configurable because it is configurable in Filebeat.

**Comment by kvch** [comment]
@ph are the errors in the E2E tests related to my change?

**Comment by ph** [comment]
@kvch I do not think they are 

```
[2022-01-28T13:55:52.855Z] WARN[2022-01-28T13:55:52Z] The Elastic artifacts API is not available yet  artifact=elastic-agent artifactName=elastic-agent-8.1.0-SNAPSHOT-linux-x86_64.tar.gz elapsedTime=52.96205476s error="GET request failed with 404" retry=15 statusEndpoint="https://artifacts-api.elastic.co/v1/search/8.1.0-dbc834fd/elastic-agent?x-elastic-no-kpi=true" version=8.1.0-dbc834fd
```

I see the unified build failed on master. https://internal-ci.elastic.co/job/elastic+unified-release+master+distributed-snapshot-multijob-master/1059/

File: x-pack/elastic-agent/_meta/config/common.reference.p2.yml.tmpl
Comment (ph) [reply x-pack/elastic-agent/_meta/config/common.reference.p2.yml.tmpl:138]: If possible, let's simplify. Sometimes less is more.

Comment (kvch) [reply x-pack/elastic-agent/_meta/config/common.reference.p2.yml.tmpl:138]: So if I understood correctly, you do not want a configuration option at all, but rather load files automatically if they are placed under a predefined directory e.g. `{path.config}/inputs.d`, right?

Comment (kvch) [reply x-pack/elastic-agent/_meta/config/common.reference.p2.yml.tmpl]: I hardcoded the glob of the external configurations to `{path.config}/inputs.d/*.yml`.

Comment (ruflin) [reply x-pack/elastic-agent/_meta/config/common.reference.p2.yml.tmpl]: Exactly. Lets remove as many config options as we can at first and only add them if we have a strong use case for these.

**Comment by ruflin** [comment]
Lets make sure we test this also as part of the e2e test framework to have some integration tests for it. 

@kvch In the Beats days, I remember there were always discussions around using `inputs` (in this case) in the inputs.d/*.yml files vs not. No strong preference on either but it would be nice if users use the wrong config, a clear message shows how it should be or link to the docs.

**Comment by kvch** [comment]
> Lets make sure we test this also as part of the e2e test framework to have some integration tests for it.

:+1:

> @kvch In the Beats days, I remember there were always discussions around using `inputs` (in this case) in the inputs.d/*.yml files vs not. No strong preference on either but it would be nice if users use the wrong config, a clear message shows how it should be or link to the docs.

I have somewhat strong preference because otherwise the file cannot be parsed. I can probably workaround the problem but I rather keep it this way. If the users miss the `inputs` line, Agent will report the error.

File: x-pack/elastic-agent/pkg/config/loader_test.go
Comment (kvch) [reply x-pack/elastic-agent/pkg/config/loader_test.go:116]: What do you mean by conflict? Setting the same ID?

**Comment by ruflin** [comment]
> I have somewhat strong preference because otherwise the file cannot be parsed. I can probably workaround the problem but I rather keep it this way. If the users miss the inputs line, Agent will report the error.

My strong preference is to make sure only 1 way works :-) Lets make sure the error that is shown to the user is very descriptive.

File: x-pack/elastic-agent/pkg/config/loader_test.go
Comment (kvch) [reply x-pack/elastic-agent/pkg/config/loader_test.go:116]: > Do we need a test case where the external and standalone configurations both specify a logs-system.syslog-* datastream for example?

In this step of configuration loading, it is irrelevant, we do not need to test it. Parsing the `inputs` configuration is done separately in `pkg/composable`. In `pkg/config` I only do an initial sanity check to make sure that there are inputs in external configuration files.

> what happens if the following is duplicated between two configurations and the other fields do not match? For example if the use_output field were different.

The same thing happens if there is a conflict in the external and standalone configuration and there are multiple inputs with the same id in the standalone config because we are merging the configs. `use_output` is an odd one out. Agent starts separate Beats for every output, because each Beat can only send to one output. Hence, there is no problem because the ID in a given Beat instance is still unique.

However, if any other field is different, it means that there are inputs in the configuration of the same Beat instance that have the same ID. In Beats ID is only used to determine which state belongs to which input if the user is collecting from the same file/journal from multiple inputs. Setting the same ID for multiple inputs leads to undefined behaviour in those that require position tracking (e.g. `journald`, `log` or `filestream`) and the same sources are configured (e.g. reading from the same file or journal).

This configurations can work without issues:
```yaml
- type: filestream
  id: my-id
  paths: [/path/to/file]
- type: filestream
  id: my-id
  paths: [/path/to/other/file]
```

Problematic configuration:
```yaml
- type: filestream
  id: my-id
  paths: [/path/to/file]
- type: filestream
  id: my-id
  paths: [/path/to/file]
```

Unfortunately, at the moment we do not check if the IDs are unique in Beats. We just tell users to set unique IDs in the documentation. We could probably add a check to be more user friendly. I am find with addressing it, but in a separate follow-up PR.

Comment (cmacknz) [reply x-pack/elastic-agent/pkg/config/loader_test.go:116]: Thanks for the great explanation. 

If there are configurations that can lead to incorrect behaviour, then we should prevent them. I agree that can be a follow up PR, or even an issue in the backlog depending on how much work it would be to implement.

I'll mark this as resolved.

**Comment by cmacknz** [comment]
Looks good to me, but I'll wait to see if someone with more knowledge of the agent configuration files can review this before approving.

**Comment by kvch** [comment]
e2e tests require https://github.com/elastic/e2e-testing/pull/2064

**Review by blakerouse — APPROVED** [review]
Overall the code looks good.

I do want to confirm the behavior that I am reading in the code. It seems this is only used when in standalone mode (which I am glad to see). But in the description of the PR it describes that it allows usage of infrastructure as code instead of Fleet. How does this help solve that, when a single elastic-agent.yml also did the same thing. The single `elastic-agent.yml` could be committed in Git as well. Just trying to understand the reason for adding this.

**Comment by kvch** [comment]
> But in the description of the PR it describes that it allows usage of infrastructure as code instead of Fleet. How does this help solve that, when a single elastic-agent.yml also did the same thing. The single `elastic-agent.yml` could be committed in Git as well. Just trying to understand the reason for adding this.

When a separate, dedicated team manages Elastic Agent, they might want to limit the access to a single `elastic-agent.yml` file. If we provide a folder from where Elastic Agent can load input configurations, the Agent manager team can allow other teams in their org to put their own input configuration there without giving access to the whole infrastructure.

**Comment by kvch** [comment]
Ooops, I merged it without a changelog entry. I will open a follow-up PR.

**Comment by kvch** [comment]
Changelog: https://github.com/elastic/beats/pull/30144

**Review by ph — COMMENTED** [review]
LGTM, after merge.

**Comment by kvch** [comment]
Documentation: https://github.com/elastic/observability-docs/pull/1515

**Pull request merged**
